hubmap/dataportal/management/commands/_metadata_import_utils.py

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env python3
+import base64
+import json
+from binascii import unhexlify
+from collections import Mapping, Sequence, defaultdict
+from hashlib import sha256
+from math import isnan
+from pathlib import Path
+from pprint import pprint  # noqa: F401
+from typing import Any, Dict, Iterable, List, Union
+
+import networkx as nx
+import pandas as pd
+from elasticsearch import Elasticsearch
+from frozendict import frozendict
+from neo4j import Driver as Neo4jDriver
+from neo4j import GraphDatabase
+from pymongo import MongoClient
+
+
+def infinite_defaultdict():
+    return defaultdict(infinite_defaultdict)
+
+
+def value_adjust(row: Iterable[Any]) -> Iterable[Any]:
+    """
+    :param row: Iterable of heterogeneous data, probably a `pd.Series`
+      in which missing values are represented as `math.nan`
+    :return: Iterable of the same data, with `nan` values replaced
+      with `None` (since `nan` values are a pain to work with in this
+      context: they don't compare equal to themselves, etc.)
+    """
+    for item in row:
+        if isinstance(item, float) and isnan(item):
+            yield None
+        else:
+            yield item
+
+
+def freeze(data):
+    """
+    :param data: Anything serializable as JSON
+    :return: An immutable and hashable version of `data`. Mappings are
+      replaced by `frozendict` instances, sequences are replaced by tuples,
+      and everything else is returned unchanged.
+    """
+    if isinstance(data, Mapping):
+        return frozendict(
+            (key, freeze(value))
+            for key, value in data.items()
+        )
+    elif isinstance(data, Sequence) and not isinstance(data, str):
+        return tuple(freeze(value) for value in data)
+    return data
+
+
+def unfreeze(data):
+    """
+    Inverse of `freeze` defined above. In our usage, we don't actually care
+    about whether the data is mutable; we just care that this can be serialized
+    as JSON. (`frozendict` instances are not.)
+    :param data:
+    :return:
+    """
+    if isinstance(data, Mapping):
+        return dict(
+            (key, unfreeze(value))
+            for key, value in data.items()
+        )
+    elif isinstance(data, Sequence) and not isinstance(data, str):
+        return [unfreeze(value) for value in data]
+    return data
+
+
+def dotted_keys_to_nested_dict(mapping: Dict[str, Any]):
+    """
+    :param mapping: A flat dict with keys as dotted strings, e.g.
+      "donor_organism.mouse_specific.strain.text", and arbitrary values
+    :return: A nested representation of that dictionary, with dotted
+      keys split into sub-dicts.
+
+    >>> d = { \
+        'a.b.c': 1, \
+        'a.b.d': 'value', \
+        'a.b.e': 5, \
+        'a.f': 'another', \
+        'g.h': None, \
+    }
+    >>> pprint(dotted_keys_to_nested_dict(d))
```

Comment:
Not sure why pprint is needed?

Reply:
I originally wanted a multi-line display of the dict, but `pprint` uses a single line if the output is within a certain width. I'll remove that.